polkadot/node/network/availability-recovery/src/task/strategy/mod.rs

```diff
@@ -0,0 +1,1534 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Recovery strategies.
+
+mod chunks;
+mod full;
+mod systematic;
+
+pub use self::{
+	chunks::{FetchChunks, FetchChunksParams},
+	full::{FetchFull, FetchFullParams},
+	systematic::{FetchSystematicChunks, FetchSystematicChunksParams},
+};
+use crate::{
+	futures_undead::FuturesUndead, ErasureTask, PostRecoveryCheck, RecoveryParams, LOG_TARGET,
+};
+
+use futures::{channel::oneshot, SinkExt};
+use parity_scale_codec::Decode;
+use polkadot_erasure_coding::branch_hash;
+#[cfg(not(test))]
+use polkadot_node_network_protocol::request_response::CHUNK_REQUEST_TIMEOUT;
+use polkadot_node_network_protocol::request_response::{
+	self as req_res, outgoing::RequestError, OutgoingRequest, Recipient, Requests,
+};
+use polkadot_node_primitives::{AvailableData, ErasureChunk};
+use polkadot_node_subsystem::{
+	messages::{AvailabilityStoreMessage, NetworkBridgeTxMessage},
+	overseer, RecoveryError,
+};
+use polkadot_primitives::{AuthorityDiscoveryId, BlakeTwo256, ChunkIndex, HashT, ValidatorIndex};
+use sc_network::{IfDisconnected, OutboundFailure, RequestFailure};
+use std::{
+	collections::{BTreeMap, HashMap, VecDeque},
+	time::Duration,
+};
+
+// How many parallel chunk fetching requests should be running at once.
+const N_PARALLEL: usize = 50;
```

Comment:
I remember asking @eskimor about this. I think it's just an estimate

Reply:
Indeed. There should be some limit otherwise I am pretty sure it will harm performance. I never measured the actual sweet spot though. This is a tunable variable.

Reply:
We can easily bisect within subsystem benchmarks to find the sweet spot. However, my own expectation is that we don't need this limit and `1/3 * n_validators` will work better.